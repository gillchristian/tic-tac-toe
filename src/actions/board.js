import { createAction } from 'redux-actions'

export const CLEAR_BOARD = 'board:clear'

// clearBoard :: () -> Redux.Action
export const clearBoard = createAction(CLEAR_BOARD)

