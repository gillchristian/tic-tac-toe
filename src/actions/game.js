import { createAction } from 'redux-actions'

export const ADD_MARK = 'game:add:mark'

// addMark :: Mark -> Int -> Redux.Action
export const addMark = createAction(
  ADD_MARK,
  (mark, position) => ({mark, position})
)

