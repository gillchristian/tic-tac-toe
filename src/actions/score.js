import { createAction } from 'redux-actions'

export const SCORE_INCREMENT = 'score:increment'

// incrementScore :: Mark -> Redux.Action
export const incrementScore = createAction(SCORE_INCREMENT)

