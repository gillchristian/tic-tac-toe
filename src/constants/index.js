export const O = '●'
export const X = '✖'
export const s = ''

export const RESULTS = [
  /1..1..1../,
  /.1..1..1./,
  /..1..1..1/,
  /111....../,
  /...111.../,
  /......111/,
  /1...1...1/,
  /..1.1.1../
]

