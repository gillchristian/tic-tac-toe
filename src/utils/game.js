import R from 'ramda'

import { X, O, s, RESULTS } from '../constants/'

// Mark :: O|X

// isO :: Mark -> Bool
export const isO = R.equals(O)
// isX :: Mark -> Bool
export const isX = R.equals(X)

// isMark :: Mark -> Bool
export const isMark = R.either(isO, isX)

// invert :: Mark -> Mark
export const invert = R.cond([
  [isO, R.always(X)],
  [isX, R.always(O)]
])

// genericCells :: Mark -> (Char -> Char)
const genericCells = (m) => R.cond([
  [R.equals(s), R.always('.')],
  [R.equals(m), R.always('1')],
  [R.T,         R.identity]
])

// genericBoard :: Mark -> ([Char] -> String)
const genericBoard = R.compose(R.map, genericCells)

//serializeOBoard :: [Char] -> String
const serializeOBoard = R.compose(R.join(''), genericBoard(O))
//serializeXBoard :: [Char] -> String
const serializeXBoard = R.compose(R.join(''), genericBoard(X))

// pickSerializer :: Mark -> ([Char] -> Char)
const pickSerializer = R.cond([
  [isO, R.always(serializeOBoard)],
  [isX, R.always(serializeXBoard)]
])

// serialize :: Mark -> [Char] -> String
export const serialize = (mark, board) => pickSerializer(mark)(board)

// checkWinner :: Mark -> [Char] -> Bool
export const checkWinner = R.curry((mark, board) => R.any(
  R.test(R.__, serialize(mark, board)),
  RESULTS
))

