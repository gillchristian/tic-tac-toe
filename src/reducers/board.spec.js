import reducer, * as b from './board'

import { X, O } from '../constants'
import { addMark } from '../actions/game'

describe('board', () => {
  describe('reducer', () => {

    describe('ADD_MARK', () => {
      it('adds the provided mark on the provided position', () => {
        let actual = reducer(b.BOARD_STATE, addMark(X, 0))
        expect(actual).toMatchSnapshot()

        actual = reducer(actual, addMark(O, 1))
        expect(actual).toMatchSnapshot()

        actual = reducer(actual, addMark(X, 2))
        expect(actual).toMatchSnapshot()

        actual = reducer(actual, addMark(O, 3))
        expect(actual).toMatchSnapshot()

        actual = reducer(actual, addMark(X, 4))
        expect(actual).toMatchSnapshot()

        actual = reducer(actual, addMark(O, 5))
        expect(actual).toMatchSnapshot()

        actual = reducer(actual, addMark(X, 6))
        expect(actual).toMatchSnapshot()

        actual = reducer(actual, addMark(O, 7))
        expect(actual).toMatchSnapshot()

        actual = reducer(actual, addMark(X, 8))
        expect(actual).toMatchSnapshot()
      })
    })

  })

  describe('selector', () => {
    const STATE = { board: b.BOARD_STATE }

    describe('getBoard', () => {
      it('gets the board state space', () => {
        expect(b.getBoard(STATE)).toEqual(b.BOARD_STATE)
      })
    })

  })
})

