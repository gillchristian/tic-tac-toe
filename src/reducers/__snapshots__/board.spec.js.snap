exports[`board reducer ADD_MARK adds the provided mark on the provided position 1`] = `
Array [
  "✖",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`board reducer ADD_MARK adds the provided mark on the provided position 2`] = `
Array [
  "✖",
  "●",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`board reducer ADD_MARK adds the provided mark on the provided position 3`] = `
Array [
  "✖",
  "●",
  "✖",
  "",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`board reducer ADD_MARK adds the provided mark on the provided position 4`] = `
Array [
  "✖",
  "●",
  "✖",
  "●",
  "",
  "",
  "",
  "",
  "",
]
`;

exports[`board reducer ADD_MARK adds the provided mark on the provided position 5`] = `
Array [
  "✖",
  "●",
  "✖",
  "●",
  "✖",
  "",
  "",
  "",
  "",
]
`;

exports[`board reducer ADD_MARK adds the provided mark on the provided position 6`] = `
Array [
  "✖",
  "●",
  "✖",
  "●",
  "✖",
  "●",
  "",
  "",
  "",
]
`;

exports[`board reducer ADD_MARK adds the provided mark on the provided position 7`] = `
Array [
  "✖",
  "●",
  "✖",
  "●",
  "✖",
  "●",
  "✖",
  "",
  "",
]
`;

exports[`board reducer ADD_MARK adds the provided mark on the provided position 8`] = `
Array [
  "✖",
  "●",
  "✖",
  "●",
  "✖",
  "●",
  "✖",
  "●",
  "",
]
`;

exports[`board reducer ADD_MARK adds the provided mark on the provided position 9`] = `
Array [
  "✖",
  "●",
  "✖",
  "●",
  "✖",
  "●",
  "✖",
  "●",
  "✖",
]
`;
